### ISSUE TITLE
gem can not be build from verifiable sources

### ISSUE THREAD

**Issue by dvzrv** [svenfuchs/test_declarative#19]
Hi! I'm currently packaging ruby-test_declarative for Arch Linux.
I realized, that the test_declarative gem can only be built from the git repository, but not from a source tarball, [because of how the .gemspec declares the files](https://github.com/svenfuchs/test_declarative/blob/master/test_declarative.gemspec#L14).

This is quite unfortunate, as it makes it easy to do a [supply-chain-attack](https://en.wikipedia.org/wiki/Supply_chain_attack) (by pushing malicious code and changing or adding a tag):
An unverifiable tag points to an unverifiable commit (both are not cryptographically signed). The checksum of the sources can not be verified either (as building from a tarball is impossible). It is only possible to clone a tag (or rather lock to a specific commit, which introduces overhead for packagers).

It would be great, if this was changed.

Thank you!

**Comment by junaruga** [comment]
> I realized, that the test_declarative gem can only be built from the git repository, but not from a source tarball, because of how the .gemspec declares the files.

Apart from the supply-chain-attack, the current situation in upstream prevented you to do packaging ruby-test_declarative for Arch Linux, right?

Where is the the source tarball? Why not create the package from the gem archive file?

```
$ gem fetch test_declarative
Fetching test_declarative-0.0.6.gem
Downloaded test_declarative-0.0.6

$ gem unpack test_declarative-0.0.6.gem

$ patch ...

$ gem spec test_declarative-0.0.6.gem -l --ruby > test_declarative.gemspec

$ gem build test_declarative.gemspec
```

On Fedora project, it does not prevent us to do packaging like this.

https://src.fedoraproject.org/rpms/rubygem-test_declarative/blob/master/f/rubygem-test_declarative.spec
https://kojipkgs.fedoraproject.org//packages/rubygem-test_declarative/0.0.5/16.fc31/data/logs/noarch/build.log
> + gem install -V --local --build-root . --force --document=ri,rdoc /builddir/build/SOURCES/test_declarative-0.0.5.gem

> This is quite unfortunate, as it makes it easy to do a supply-chain-attack (by pushing malicious code and changing or adding a tag):

This can be the alternative.

https://github.com/rack/rack/blob/master/rack.gemspec#L20-L21

**Comment by dvzrv** [comment]
> Apart from the supply-chain-attack, the current situation in upstream prevented you to do packaging ruby-test_declarative for Arch Linux, right?

No, as I'm currently locking to the specific commit the tag points to. However, this is needlessly cumbersome.

A patch of the gemspec is of course an alternative (also for packaging) and should actually be considered the canonical way of fixing this. Thanks for the pointer @junaruga!

**Comment by junaruga** [comment]
> No, as I'm currently locking to the specific commit the tag points to. However, this is needlessly cumbersome.
>
> A patch of the gemspec is of course an alternative (also for packaging) and should actually be considered the canonical way of fixing this. Thanks for the pointer @junaruga!

Alright. Is it possible to send the pull-request to master branch to fix this issue?
